eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/LazyIterate.java

```diff
@@ -196,4 +198,14 @@ private LazyIterate()
     {
         return new TapIterable<>(iterable, procedure);
     }
+
+    public static <A, B> LazyIterable<Pair<A, B>> cartesianProduct(Iterable<A> iterable1, Iterable<B> iterable2)
+    {
+        return LazyIterate.cartesianProduct(iterable1, iterable2, Tuples::pair);
+    }
+
+    public static <A, B, C> LazyIterable<C> cartesianProduct(Iterable<A> iterable1, Iterable<B> iterable2, Function2<A, B, C> function)
+    {
+        return LazyIterate.flatCollect(iterable1, first -> LazyIterate.collect(iterable2, second -> function.value(first, second)));
```

Comment:
@motlin so what is the conclusion on this? Is it ok like this or should we stop writing procedural code and introduce a `cartesianProduct` (or equivalent name) method on collections interfaces in an object-oriented way to handle each implementation with respect to the concerned type?

I believe this is out of scope of this PR in the latter case... `LazyIterate.cartesianProduct` is an helper method just to simplify life of people because writing this flatCollect/collect/apply is common and boring to write.

Honestly having a `cartesianProduct` for `Bag`s or `Set`s does not interest me at all, I just want to ease my life because I often have to cross join multiple collections.

So should we continue with this? If yes I will add something about the complexity in the javadoc at least. I can't make this decision, so I'm referring to you all :)

Reply:
I don't think those are the only two choices. The implementation for Sets returns a `FlatCollectIterable`. That's fine for sets, because they're already unique, and no additional work is required than the work done inside the lambda.

I think you should considering implementing LazyIterate.cartesianProduct() by "simply" returning a new type `CartesianProductIterable`. That would give you a new type with methods like toSet() and toBag() that you could optimize. For sets, the optimization could be to delegate to distinct() before proceeding. For bags, you could delegate to forEachWithOccurrences if available, or toBag followed by forEachWithOccurrences if not. It's a little tricky, and a lot of work for an obscure method, so I'd be open to a simpler implementation plus filing a follow-up issue for the optimization work. What do you think?

Reply:
The `CartesianProductIterable` could also be add as an optimization later, as long as the method returns `LazyIterable`, which it does currently.